### ISSUE TITLE
Replace a source file string in realtime

### ISSUE THREAD

**Issue by ghost** [microsoft/language-server-protocol#1638]
Hi, not an issue but a newbie question. If I type `\alpha` in the editor, can a LS find it out and replace it with the respective greek letter without any further user action more complicated than, say, hitting a tab?

**Comment by mickaelistria** [comment]
This can probably be properly achieved with [onTypeFormatting](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_onTypeFormatting)

**Comment by dbaeumer** [comment]
Although `onTypeFormatting` will be able to do this I would not recommend this. It is not enforced but formatting should usually only manipulate white space and if the formatting changes stuff around the cursor the client might drop it sine it can't be reconciled.

So the answer is it can not be done reliable right now since we have no support for proper reconciling in these situaitons.

**Comment by mickaelistria** [comment]
Another approach is to detects `\alpha` in the code and  on `documentChanged` and react bu sending an `applyEdit` with a TextEdit to replace the range of `\alpha` by `ɑ`.

**Comment by ghost** [comment]
`applyEdit` would work but I wonder about the complexity of scanning the whole document after its every single modification. To solve the problem, the server might perform such a scan after `documentChanged` but only if one second has passed since the previous scan. But then, a "skipped" `documentChanged` might turn out to be the last one before a long pause, so in effect the text conversion in question would be pending for a long time which in turn might look like something is wrong. This could be prevented by an asynchronous `applyEdit` generated by the server a second into the pause. But are such asynchronous operations possible?

**Comment by mickaelistria** [comment]
> This could be prevented by an asynchronous applyEdit generated by the server a second into the pause. But are such asynchronous operations possible?

Yes. Many language server do actually implement the document scanning that way: they accumulate changes as they come on only trigger analysis (and reaction such as publishDiagnostics or applyEdits), when users seems to have "paused" typing for enough time, typically 1 or 2 seconds.

**Comment by ghost** [comment]
So LSP has probably all the functionality I would need here.

**Comment by dbaeumer** [comment]
But please be aware that apply edit is version based. This means that if the user kept typing the version of the document will have been advanced and the edit will be rejected. The apply edit doesn't allow for concurrent editing either.

**Comment by ghost** [comment]
But if the server sees that its modification has been rejected, I guess that it can wait for a user inactivity and then repeat asynchronously the read new version/analyze/applyEdit cycle until it works, as mickaelistria suggests?

**Comment by dbaeumer** [comment]
Yes, that would be possible
